man/edit_tape.1
.TH edit_tape "1" "December 2013" "mhvtl 1.5" "User Commands"
.SH NAME
edit_tape \- Utility to update meta data virtual media for the VTL.
.SH SYNOPSIS
.B edit_tape
.B [ \-h ]
.B [ \fIoptions \fR]
.B ...
.SH DESCRIPTION
.\" Add any additional description here
.PP
Create and initialise new media for the Virtual Tape Library. Media is created in the
/opt/vtl/ directory. This path is hard coded.
.SH OPTIONS
.TP
\fB\-h\fR
display this help and exit
.TP
\fB\-m PCL\fR
where PCL is the Physical Cartridge Label (barcode).
.TP
\fB\-w on|off\fR
Turn on|off media write-protect flag
.TP
\fB\-s size\fR
where size is the capacity of the virtual media - Size is defined in 'megabytes'.
.TP
\fB\-t type\fR
The media 'type' which can be: "data" , "WORM" (Write Once Read Many) or
"clean" (cleaning cartridge)
.TP
\fB\-d density\fR
Media density. Can be one of LTO1, LTO2, LTO3, LTO4, SDLT1, SDLT2, SDLT3, SDLT4, AIT1, AIT2, AIT3,
AIT4, T10KA, T10Kb, T10KC, 9840A, 9840B, 9840C, 9840D, 9940A, 9940B, J1A, E05 and E06

The 'J1A, E05 and E06' media densities refer to the IBM 03592 media types.

.SH AUTHOR
Written by Mark Harvey
.SH BUGS
Needs to be made user friendly.
.SH "REPORTING BUGS"
Report bugs to <markh794@gmail.com> <mark_harvey@symantec.com>.
.SH COPYRIGHT
Copyright \(co 2005 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.BR build_library_config(1),
.BR make_vtl_media(1),
.BR library_contents(5),
.BR mhvtl(1),
.BR vtlcmd(1),
.BR vtllibrary(1),
.BR vtltape(1)
.BR mktape(1)


ChangeLog
* Sun Oct 20 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-10
- Various cleanups due to static code check 'smatch'
- Fix TapeAlert bit offsets
- Return bit/byte offset for sense INVALID FIELD IN CDB/PARAMS
- Fix returned data size when 0 data is requested (kernel module fix)
- SSC - Fix op code 0Fh (read reverse) incorrectly set to 'reserve'
- SMC - Ability to keep library media change persistant across restarts
- SSC - Implement 'read media serial number' op code
- Fix 'make_vtl_media' script to understand IBM 03592 media type

* Thu Aug 29 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-9
- SSC -> MODE SELECT overhaul.
- Sanitize delay values introduced in 1.4-8
- daemons now open /dev/mhvtlXX nodes in exclusive mode
- Code cleanup based on Valgrind - memory leaks and unused file handles
- Correctly set SCSI revision field for T10000B
- SMC -> Adjust behaviour if 'Scalar' library defined
 set 'BarC' field in inquiry page and
 pad "Full Firmware Revision" field with compile time/date

* Sat Jun 29 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-8
- Add additional fields in MAM for multi-partition support capability
- MODE SELECT fixes
- Solaris 11.1 st mode select
- Make virtual media more resilient against disk full conditions
- Add additional error checking between kernel and userspace ioctl
- Add ability to introduce delays in load/unload/position/thread/rewind op codes
- Add 16byte CDB locate command support
- Minor kernel module tweeks - increase timeout & error checking on ioctl()

* Fri Mar 22 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-7
- Add DLT7000/DLT8000 emulation
- Record 'last load' data in MAM
- New utility 'edit_tape' which will allow modificiation of media 'meta data'
- Prevent and log possibility of READ ELEMENT STATUS overflowing data buffer
- Set 'POWERON-RESET' if library re-configured via HUP signal
- Update 'minor' number from 'unsigned char' to 'unsigned int' which removes
 max number of device nodes at 256
- Rework READ ELEMENT STATUS - Simplify code and correct bug where incorrect
 byte count returned.
- Add 'support' for specifying 9x40 and DLT media via barcode suffix
- Update kernel module to prevent overflow if DEF_MAX_MINOR_NO reached.
- Update kernel module to allow up to 1024 minor numbers
- Update kernel module vers to 0.18.12, vers date 20130310-0
- off-by-one fix of compression mode page
- Fix 'READ POSITION' ssc op code
- Helper scripts for NetBackup added
 - drive_stats.pl (parse 'scsi_command' output and email on tape unload)
 - vlt_endeject_notify.pl (NetBackup Vault helper)

* Sat Jan 31 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-6
- Return correct TAPE CAPACITY log page
- Fix overflow TAPE CAPACITY log page
- Fix potential vtltape segfault when 'block position back'
- New utility 'edit_tape' to update virtual media metadata

* Sat Jan 12 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-5
- Added support for LTO-6
- Fix EOM handling
- Fix data returned in TAPE CAPACITY log page

* Thu Sep 13 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-4
- Fix a build issue on older (SLES9) linux distributions
- Silence a compiler warning about uninitialised var 'month' in quantum PM
- Note: 1.4-2 & 1.4-3 never made it as released packages.

* Wed Sep 12 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-3
- Add 'performance' counter to vtllibrary
- Fix off-by-one error in LOG SELECT
- Implement LOG reset via PCR bit in LOG SELECT

* Sat Sep 01 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-2
- Add 'performance' counter to each SCSI op code logged

* Wed Aug 08 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-1
- Bug fix:
  Fix dump_tape to understand new library subdirectory format
  Fix vtlcmd to understand new library subdirectory format

* Wed Aug 01 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-0
- Allow 'default' personality module to load all media R/W
- Add IBM half-hight INQUIRY string match
- vtltape: Update log entry regarding media capacity depending on capabilities
- vtltape: Make the backoff algrithm value configurable
- Add SDLT600 & SDLT320 personality module
- mhvtl.ko: Silence install 'depmod' error
- Move standard inquiry info into memory structure
- Default media belonging to a library in a subdir
- Remove kmod-mhvtl.spec
- Rename 'mhvtl' package as 'mhvtl-utils'
- Use ELRepo spec file
- Move 'make install' from RPM to Makefile

* Fri Jun 15 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.3-0
- Move 'on disk' unique functions into its own shared lib (libvtlcart.so)
- HP Ultrium VPD page data update to match HP documentation
- IBM Ultrium VPD page 0xC0 & 0xC1 update to match IBM documentation
- Fix SPIN return status
- Add helper functions for moving media within library
  (in prep for library partition manager)
- Add ability to set/clear APPEND ONLY status from user (vtlcmd)

* Wed Apr 04 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.2-3
- Fix segfault in AIT4 when attempting to access Security Protocol IN/OUT
- No need to fail in virtual media parent directory exists
- Add more descriptive debug messages for log/mode page lookups

* Thu Mar 22 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.2-2
- Changed defaults so fifo is not enabled. If fifo not used, it will
  block writing process (i.e. hang daemons)
- Improved MODE SELECT / SENSE to support 'changeable field bitmaps
- Started implementing ALLOW OVERWRITE support (incomplete)

usr/edit_tape.c
/*
 * edit_tape is portion of the mhVTL package.
 *
 * The vtl package consists of:
 *   a kernel module (vlt.ko) - Currently on 2.6.x Linux kernel support.
 *   SCSI target daemons for both SMC and SSC devices.
 *
 * Copyright (C) 2005 - 2013 Mark Harvey       markh794@gmail.com
 *                                          mark_harvey@symantec.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <syslog.h>
#include <string.h>
#include <time.h>
#include <inttypes.h>
#include "be_byteshift.h"
#include "list.h"
#include "vtl_common.h"
#include "vtltape.h"
#include "vtllib.h"
#include "logging.h"

#if defined _LARGEFILE64_SOURCE
static void *largefile_support = "large file support";
#else
static void *largefile_support = "No largefile support";
#endif

/* The following variables are needed for the MHVTL_DBG() macro to work. */

char vtl_driver_name[] = "edit_tape";
int verbose;
int debug;
int wp;	/* Write protect flag */
long my_id;
extern char home_directory[HOME_DIR_PATH_SZ + 1];

#define WRITE_PROTECT_OFF 1
#define WRITE_PROTECT_ON  2

void usage(char *progname)
{
	printf("Usage: %s -l lib -m PCL [-s size] [-t type] [-d density]"
		" [-w on|off]\n",
					progname);
	printf("       Where 'size' is in Megabytes\n");
	printf("             'lib' is Library number\n");
	printf("             'type' is data | clean | WORM\n");
	printf("             'PCL' is Physical Cartridge Label (barcode)\n");
	printf("             '-w on|off' enable/disable write protect flag\n");
	printf("             'density' can be on of the following:\n");
	printf("           AIT1     AIT2     AIT3     AIT4\n");
	printf("           DDS1     DDS2     DDS3     DDS4\n");
	printf("           DLT3     DLT4\n");
	printf("           SDLT1    SDLT220  SDLT320  SDLT600\n");
	printf("           LTO1     LTO2     LTO3     LTO4\n");
	printf("           LTO5     LTO6\n");
	printf("           T10KA    T10KB    T10KC\n");
	printf("           9840A    9840B    9840C    9840D\n");
	printf("           9940A    9940B\n");
	printf("           J1A      E05      E06\n\n");
}

int main(int argc, char *argv[])
{
	unsigned char sam_stat;
	char *progname = argv[0];
	char *pcl = NULL;
	char *mediaType = NULL;
	char *mediaCapacity = NULL;
	char *density = NULL;
	uint64_t size;
	struct MAM new_mam;
	char *lib = NULL;
	int libno = 0;
	int indx;
	int rc;
	char *config = MHVTL_CONFIG_PATH"/device.conf";
	FILE *conf;
	char *b;	/* Read from file into this buffer */
	char *s;	/* Somewhere for sscanf to store results */

	if (sizeof(struct MAM) != 1024) {
		printf("Structure of MAM incorrect size: %d\n",
						(int)sizeof(struct MAM));
		exit(2);
	}

	if (argc < 2) {
		usage(progname);
		exit(1);
	}

	debug = 0;
	my_id = 0;
	verbose = 0;
	wp = 0;

	while (argc > 0) {
		if (argv[0][0] == '-') {
			switch (argv[0][1]) {
			case 'd':
				if (argc > 1)
					density = argv[1];
				break;
			case 'l':
				if (argc > 1) {
					lib = argv[1];
				} else {
					puts("    More args needed for -l\n");
					exit(1);
				}
				break;
			case 'm':
				if (argc > 1) {
					pcl = argv[1];
				} else {
					puts("    More args needed for -m\n");
					exit(1);
				}
				break;
			case 's':
				if (argc > 1) {
					mediaCapacity = argv[1];
				} else {
					puts("    More args needed for -s\n");
					exit(1);
				}
				break;
			case 't':
				if (argc > 1) {
					mediaType = argv[1];
				} else {
					puts("    More args needed for -t\n");
					exit(1);
				}
				break;
			case 'V':
				printf("%s: version %s\n%s\n\n",
						progname, MHVTL_VERSION,
						(char *)largefile_support);
				break;
			case 'v':
				verbose++;
				break;
			case 'w':
				if (argc > 1) {
					if (!strncasecmp("yes", argv[1], 3))
						wp = WRITE_PROTECT_ON;
					else if (!strncasecmp("on", argv[1], 3))
						wp = WRITE_PROTECT_ON;
					else
						wp = WRITE_PROTECT_OFF;
				} else {
					puts("    More args needed for -m\n");
					exit(1);
				}
				break;
			}
		}
		argv++;
		argc--;
	}

	if (pcl == NULL) {
		printf("Please supply a barcode (-b barcode)\n\n");
		usage(progname);
		exit(1);
	}

	conf = fopen(config , "r");
	if (!conf) {
		printf("Can not open config file %s : %s", config,
					strerror(errno));
		perror("Can not open config file");
		exit(1);
	}
	s = zalloc(MALLOC_SZ);
	if (!s) {
		perror("Could not allocate memory");
		exit(1);
	}
	b = zalloc(MALLOC_SZ);
	if (!b) {
		perror("Could not allocate memory");
		exit(1);
	}

	rc = ENOENT;

	if (lib) {
		sscanf(lib, "%d", &libno);
		printf("Looking for PCL: %s in library %d\n", pcl, libno);
		find_media_home_directory(home_directory, libno);
		rc = load_tape(pcl, &sam_stat);
	} else { /* Walk thru all defined libraries looking for media */
		while (readline(b, MALLOC_SZ, conf) != NULL) {
			if (b[0] == '#')	/* Ignore comments */
				continue;
			/* If found a library: Attempt to load media
			 * Break out of loop if found. Otherwise try next lib.
			 */
			if (sscanf(b, "Library: %d CHANNEL:", &indx)) {
				find_media_home_directory(home_directory, indx);
				rc = load_tape(pcl, &sam_stat);
				if (!rc)
					break;
			}
		}
	}

	fclose(conf);
	free(s);
	free(b);

	if (rc) {
		fprintf(stderr, "PCL %s cannot be dumped, "
				"load_tape() returned %d\n",
					pcl, rc);
		exit(1);
	}

	/* Copy media MAM into temp location */
	memcpy(&new_mam, &mam, sizeof(mam));

	size = 0L;
	if (mediaCapacity) {
		sscanf(mediaCapacity, "%" PRId64, &size);
		printf("New capacity for %s: %ldMB\n",
					pcl, (unsigned long)size);
	}

	if (mediaType) {
		if (strncmp("clean", mediaType, 5)) {
			MHVTL_DBG(1, "Setting media type to CLEAN");
			new_mam.MediumType = MEDIA_TYPE_CLEAN;
			new_mam.MediumTypeInformation = 20;
		} else if (!strncmp("WORM", mediaType, 4)) {
			MHVTL_DBG(1, "Setting media type to WORM");
			new_mam.MediumType = MEDIA_TYPE_WORM;
		} else {
			MHVTL_DBG(1, "Setting media type to DATA");
			new_mam.MediumType = MEDIA_TYPE_DATA;
		}
	}
	if (density) {
		printf("Setting density to %s\n", density);
		if (set_media_params(&new_mam, density)) {
			printf("Could not determine media density: %s\n",
					density);
			unload_tape(&sam_stat);
			exit(1);
		}
	}
	if (size) {
		put_unaligned_be64(size * 1048576, &new_mam.max_capacity);
		/* This will set incorrect value to start with but next media
		 * Usage, this will be recalculated correctly
		 */
		put_unaligned_be64(size * 1048576, &new_mam.remaining_capacity);
	}
	switch (wp) {
	case WRITE_PROTECT_ON:
		new_mam.Flags |= MAM_FLAGS_MEDIA_WRITE_PROTECT;
		printf("Setting write-protect for %s\n", pcl);
		break;
	case WRITE_PROTECT_OFF:
		new_mam.Flags &= ~MAM_FLAGS_MEDIA_WRITE_PROTECT;
		printf("Turning off write-protect for %s\n", pcl);
		break;
	}

	put_unaligned_be64(sizeof(mam.pad), &new_mam.MAMSpaceRemaining);

	memcpy(&mam, &new_mam, sizeof(mam));
	rewriteMAM(&sam_stat);
	unload_tape(&sam_stat);

	exit(0);
}
